psi4/driver/driver_nbody.py

```diff
@@ -225,65 +211,188 @@ def nbody_gufunc(func, method_string, **kwargs):
+    for r in [component_results, nbody_results]:
+        for d in r:
+            if d in dicts:
+                for var, value in r[d].items():
+                    wfn.set_variable(str(var), value)
+                    core.set_variable(str(var), value)
+
+    if metadata['ptype'] == 'gradient':
+        wfn.set_gradient(ret_ptype)
+    elif metadata['ptype'] == 'hessian':
+        wfn.set_hessian(ret_ptype)
+
+    core.set_variable("CURRENT ENERGY", nbody_results['ret_energy'])
+    wfn.set_variable("CURRENT ENERGY", nbody_results['ret_energy'])
+
+    if metadata['return_wfn']:
+        return (nbody_results['ret_ptype'], wfn)
+    else:
+        return nbody_results['ret_ptype']
+
+def build_nbody_compute_list(metadata):
+    """Generates the list of N-Body computations to be performed for a given BSSE type.
+
+    Parameters
+    ----------
+    metadata : dict of str
+        Dictionary containing N-body metadata.
+
+        Required ``'key': value`` pairs:
+        ``'bsse_type_list'``: list of str
+            List of requested BSSE treatments.  Possible values include lowercase ``'cp'``, ``'nocp'``,
+            and ``'vmfc'``.
+        ``'max_nbody'``: int
+            Maximum number of bodies to include in the N-Body treatment.
+            Possible: `max_nbody` <= `max_frag`
+            Default: `max_nbody` = `max_frag`
+        ``'max_frag'``: int
+            Number of distinct fragments comprising full molecular supersystem.
+
+    Returns
+    -------
+    metadata : dict of str
+        Dictionary containing N-body metadata.
+        
+        New ``'key': value`` pair:
+        ``'compute_dict'`` : dict of str: dict
+            Dictionary containing subdicts enumerating compute lists for each possible BSSE treatment.
+
+            Contents:
+            ``'all'``: dict of int: set
+                Set containing full list of computations required
+            ``'cp'``: dict of int: set
+                Set containing list of computations required for CP procedure
+            ``'nocp'``: dict of int: set
+                Set containing list of computations required for non-CP procedure
+            ``'vmfc_compute'``: dict of int: set
+                Set containing list of computations required for VMFC procedure
+            ``'vmfc_levels'``: dict of int: set
+                Set containing list of levels required for VMFC procedure
+    """
+    # What levels do we need?
+    metadata['nbody_range'] = range(1, metadata['max_nbody'] + 1)
+    metadata['fragment_range'] = range(1, metadata['max_frag'] + 1)
+
+    cp_compute_list = {x:set() for x in metadata['nbody_range']}
+    nocp_compute_list = {x:set() for x in metadata['nbody_range']}
+    vmfc_compute_list = {x:set() for x in metadata['nbody_range']}
+    vmfc_level_list = {x:set() for x in metadata['nbody_range']} # Need to sum something slightly different
 
     # Build up compute sets
-    if do_cp:
+    if 'cp' in metadata['bsse_type_list']:
```

Comment:
It would be good to verify the `bsse_type_list` here again. Fun with sets: `if len({'cp', 'nocp', 'vmfc'} - set(metadata['bsse_type_list'])): raise Exception()`.